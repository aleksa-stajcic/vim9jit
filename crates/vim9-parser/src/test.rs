#![cfg(test)]

use crate::*;

macro_rules! snapshot {
    ($name:tt, $path:tt) => {
        #[test]
        fn $name() {
            setup_trace();

            let contents = include_str!($path);
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path("../testdata/output/");
            settings.bind(|| {
                insta::assert_snapshot!(snapshot_parsing(contents));
            });
        }
    };
}

snapshot!(test_var, "../testdata/snapshots/simple_var.vim");
snapshot!(test_ret, "../testdata/snapshots/simple_ret.vim");
snapshot!(test_shared, "../testdata/snapshots/shared.vim");
snapshot!(test_comment, "../testdata/snapshots/comment.vim");
snapshot!(test_header, "../testdata/snapshots/header.vim");
snapshot!(test_expr, "../testdata/snapshots/expr.vim");
snapshot!(test_echo, "../testdata/snapshots/echo.vim");
snapshot!(test_scopes, "../testdata/snapshots/scopes.vim");
snapshot!(test_autocmd, "../testdata/snapshots/autocmd.vim");
snapshot!(test_array, "../testdata/snapshots/array.vim");
snapshot!(test_dict, "../testdata/snapshots/dict.vim");
snapshot!(test_if, "../testdata/snapshots/if.vim");
snapshot!(test_call, "../testdata/snapshots/call.vim");
snapshot!(test_concat, "../testdata/snapshots/concat.vim");
snapshot!(test_unpack, "../testdata/snapshots/unpack.vim");
snapshot!(test_assign, "../testdata/snapshots/assign.vim");
snapshot!(test_vimvar, "../testdata/snapshots/vimvar.vim");
snapshot!(test_busted, "../testdata/snapshots/busted.vim");
snapshot!(test_heredoc, "../testdata/snapshots/heredoc.vim");
snapshot!(test_typed_params, "../testdata/snapshots/typed_params.vim");
snapshot!(test_index, "../testdata/snapshots/index.vim");
snapshot!(test_adv_index, "../testdata/snapshots/adv_index.vim");
snapshot!(test_multiline, "../testdata/snapshots/multiline.vim");
snapshot!(test_cfilter, "../testdata/snapshots/cfilter.vim");
snapshot!(test_lambda, "../testdata/snapshots/lambda.vim");
snapshot!(test_comparisons, "../testdata/snapshots/comparisons.vim");
snapshot!(test_methods, "../testdata/snapshots/methods.vim");
snapshot!(test_eval, "../testdata/snapshots/eval.vim");
snapshot!(test_export, "../testdata/snapshots/export.vim");
snapshot!(test_import, "../testdata/snapshots/import.vim");
snapshot!(
    test_plugin_fileselect,
    "../testdata/snapshots/plugin_fileselect.vim"
);

#[test]
fn test_peek_n() {
    let input = "vim9script\nvar x = true\n";
    let lexer = new_lexer(input);
    let mut parser = Parser::new(lexer);
    assert_eq!(parser.peek_token.kind, TokenKind::EndOfLine);
    assert_eq!(parser.peek_n(0).kind, TokenKind::Identifier);
    assert_eq!(parser.peek_n(1).kind, TokenKind::EndOfLine);
    assert_eq!(parser.peek_n(2).kind, TokenKind::Identifier);
    assert_eq!(parser.peek_n(3).kind, TokenKind::Identifier);
    assert_eq!(parser.peek_n(4).kind, TokenKind::Equal);
    assert_eq!(parser.peek_n(1).kind, TokenKind::EndOfLine);
}

// TODO: Slowly but surely, we can work towards this
// snapshot!(test_matchparen, "../../shared/snapshots/matchparen.vim");
