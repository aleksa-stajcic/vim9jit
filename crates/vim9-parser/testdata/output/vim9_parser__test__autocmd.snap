---
source: crates/vim9-parser/src/test.rs
assertion_line: 29
expression: snapshot_parsing(contents)
---
[
    Vim9Script(
        Vim9ScriptCommand {
            noclear: false,
            eol: Token(EndOfLine, "\n", (0,10)->(0,10)),
        },
    ),
    Augroup(
        AugroupCommand {
            augroup: Token(Identifier, "augroup", (2,0)->(2,7)),
            augroup_name: Literal {
                token: Token(Identifier, "matchparen", (2,8)->(2,18)),
            },
            augroup_eol: Token(EndOfLine, "\n", (2,18)->(2,18)),
            body: Body {
                commands: [
                    Comment(
                        Token(Comment, "# Replace all matchparen autocommands", (3,2)->(3,39)),
                    ),
                    Autocmd(
                        AutocmdCommand {
                            autocmd: Token(Identifier, "autocmd", (4,2)->(4,9)),
                            bang: true,
                            events: [
                                Literal {
                                    token: Token(Identifier, "CursorMoved", (4,11)->(4,22)),
                                },
                                Literal {
                                    token: Token(Identifier, "CursorMovedI", (4,23)->(4,35)),
                                },
                                Literal {
                                    token: Token(Identifier, "WinEnter", (4,36)->(4,44)),
                                },
                            ],
                            pattern: Pattern(
                                [
                                    "*",
                                ],
                            ),
                            block: Block(
                                Block {
                                    open: Token(LeftBrace, "{", (4,47)->(4,48)),
                                    body: Body {
                                        commands: [
                                            NoOp(
                                                Token(EndOfLine, "\n", (4,48)->(4,48)),
                                            ),
                                            Echo(
                                                EchoCommand {
                                                    echo: Token(Identifier, "echo", (5,6)->(5,10)),
                                                    expr: String(
                                                        DoubleQuote(
                                                            "Block",
                                                        ),
                                                    ),
                                                    eol: Token(EndOfLine, "\n", (5,18)->(5,18)),
                                                },
                                            ),
                                        ],
                                    },
                                    close: Token(RightBrace, "}", (6,4)->(6,5)),
                                    eol: Token(EndOfLine, "\n", (6,5)->(6,5)),
                                },
                            ),
                        },
                    ),
                    NoOp(
                        Token(EndOfLine, "\n", (7,0)->(7,0)),
                    ),
                    Autocmd(
                        AutocmdCommand {
                            autocmd: Token(Identifier, "autocmd", (8,2)->(8,9)),
                            bang: false,
                            events: [
                                Literal {
                                    token: Token(Identifier, "WinLeave", (8,10)->(8,18)),
                                },
                            ],
                            pattern: Pattern(
                                [
                                    "*",
                                ],
                            ),
                            block: Command(
                                Echo(
                                    EchoCommand {
                                        echo: Token(Identifier, "echo", (8,21)->(8,25)),
                                        expr: String(
                                            DoubleQuote(
                                                "Command",
                                            ),
                                        ),
                                        eol: Token(EndOfLine, "\n", (8,35)->(8,35)),
                                    },
                                ),
                            ),
                        },
                    ),
                    NoOp(
                        Token(EndOfLine, "\n", (9,0)->(9,0)),
                    ),
                    Autocmd(
                        AutocmdCommand {
                            autocmd: Token(Identifier, "autocmd", (10,2)->(10,9)),
                            bang: false,
                            events: [
                                Literal {
                                    token: Token(Identifier, "WinLeave", (10,10)->(10,18)),
                                },
                            ],
                            pattern: Pattern(
                                [
                                    "*.py",
                                ],
                            ),
                            block: Command(
                                Echo(
                                    EchoCommand {
                                        echo: Token(Identifier, "echo", (10,24)->(10,28)),
                                        expr: String(
                                            DoubleQuote(
                                                "snek lang",
                                            ),
                                        ),
                                        eol: Token(EndOfLine, "\n", (10,40)->(10,40)),
                                    },
                                ),
                            ),
                        },
                    ),
                    Autocmd(
                        AutocmdCommand {
                            autocmd: Token(Identifier, "autocmd", (11,2)->(11,9)),
                            bang: false,
                            events: [
                                Literal {
                                    token: Token(Identifier, "WinLeave", (11,10)->(11,18)),
                                },
                            ],
                            pattern: Pattern(
                                [
                                    "*.py",
                                    "*.rs",
                                ],
                            ),
                            block: Command(
                                Echo(
                                    EchoCommand {
                                        echo: Token(Identifier, "echo", (11,29)->(11,33)),
                                        expr: String(
                                            DoubleQuote(
                                                "One of these is better",
                                            ),
                                        ),
                                        eol: Token(EndOfLine, "\n", (11,58)->(11,58)),
                                    },
                                ),
                            ),
                        },
                    ),
                    NoOp(
                        Token(EndOfLine, "\n", (12,0)->(12,0)),
                    ),
                ],
            },
            augroup_end: Token(Identifier, "augroup", (13,0)->(13,7)),
            augroup_end_name: Token(Identifier, "END", (13,8)->(13,11)),
            augroup_end_eol: Token(EndOfLine, "\n", (13,11)->(13,11)),
        },
    ),
]
